kelda: Replace scheduler with Kubernetes

kklin opened the pull request:
Before, we had a custom scheduler for ensuring the desired containers
were running. While it is not too difficult to implement a basic
scheduler, moving to Kubernetes will allow us to iterate more quickly on
features such as volumes since Kubernetes already implements much of the
logic for actually creating and managing volumes -- we will be able to
just focus on how to express volumes in the blueprint language.

Furthermore, this change allows us to potentially, in the future, run on
top of an existing Kubernetes cluster. This will be helpful given the
general movement in industry towards using Kubernetes.

At a high level, this commit encompasses four major changes:
1) It modifies the deployment engine to boot a Kubernetes cluster on the
Kelda machines.
2) It replaces the custom scheduler with code that translates the
containers specified by the user into deployments to be enforced by the
Kubernetes cluster.
3) It replaces the Docker network plugin with a CNI network plugin
compatible with Kubernetes.
4) It replaces Vault in favor of the built-in Kubernetes solution for
secret management. It's simpler to just use the Kubernetes solution, and
we weren't making use of any of the more advanced Vault features.

kklin commented:
@ejj I left a couple small documentation TODOs for you. You can find them by searching for "TODO Ethan". There were also some TODOs that you left from your CNI work that I don't know if you want to address.

ejj commented:
Are there any dependencies we can remove from the vendor directory?

kklin commented:
FYI @ejj this will change a bit to support the volumes/privileged stuff that was merged yesterday. Up to you how you want to review it

ejj commented on cli/command/debug.go:
```diff
@@ -170,33 +175,38 @@ func (dCmd Debug) Run() int {
 	// we might be unable to query the current containers because the master
 	// minion crashed, in which case the machine logs would be very useful.
 	var numQueryErrors int
```
Perhaps I'm reading the code wrong, but this variable doesn't appear to be used at all (other than incremented).  Not the patch to fix it of course ...

kklin replied:
It's used in the return statement

ejj commented:
FWIW, I think I would have liked to see the fuzzy lookup change be in it's own PR.  Seems worthwhile, but I don't see what it has to do with the Kube migration.

kklin commented:
I assume you meant the fuzzy lookup change to remove the argument for the public IP of the return thing. It was needed before because SSHing into a container required SSHing into the public IP of the worker, which differed according to the container. With Kube, we just always go through the leader, so that's why it's removed in this PR. I guess I could have left it unused in the Kube commit, and then removed it entirely in a later one.

ejj commented:
Ah fair enough.

ejj commented on minion/kubernetes/affinity.go:
```diff
@@ -0,0 +1,190 @@
+			handlePodAffinity(affinity, plcm.OtherContainer, plcm.Exclusive)
+		}
+
+		nodeConstraints := []struct{ key, val string }{
```
Why not use a `map[string]string`?

kklin replied:
The order needs to be consistent. I'll add a comment

ejj commented on minion/kubernetes/affinity.go:
```diff
@@ -0,0 +1,190 @@
+			}
+		}
+
+		if needsUpdate {
```
I think I would `if !needsUpdate { continue }`.  Doesn't matter much though

ejj commented on minion/kubernetes/configmap.go:
```diff
@@ -0,0 +1,122 @@
+// These config maps are referenced by the updateDeployment function to mount
+// the container's filepathToContent into the container.
+func updateConfigMaps(conn db.Conn, configMapsClient clientv1.ConfigMapInterface) (
+	noErrors bool) {
```
This seems odd, why not just return an error?

kklin replied:
It's to avoid the complexity of aggregating multiple errors. The caller wouldn't make use of the error anyways, it would just check whether it was `== nil`. What do you think?

ejj commented on minion/kubernetes/configmap.go:
```diff
@@ -0,0 +1,122 @@
+package kubernetes
```
I think this file would benefit from an overview comment.

ejj commented on minion/kubernetes/deployment.go:
```diff
@@ -0,0 +1,405 @@
+	for _, intf := range toCreate {
+		deployment := intf.(appsv1.Deployment)
+		log.WithField("deployment", deployment.Name).
+			Info("Creating deployment")
```
FWIW, this fits on the above line.

ejj commented on minion/kubernetes/deployment.go:
```diff
@@ -0,0 +1,405 @@
+	for _, intf := range toDelete {
+		deployment := intf.(appsv1.Deployment)
+		log.WithField("deployment", deployment.Name).
+			Info("Deleting deployment")
```
Same here

ejj commented on minion/kubernetes/deployment.go:
```diff
@@ -0,0 +1,405 @@
+	return deployments
+}
+
+type deploymentBuilder struct {
```
Looking at this, I don't think it's buying us very much.  I think I'd be inclined just to use a function rather than having this config object with methods.  Passing the config isn't so big a deal.

ejj commented on minion/kubernetes/deployment.go:
```diff
@@ -0,0 +1,405 @@
+
+func (builder deploymentBuilder) makeDeploymentForContainer(dbc db.Container) (
+	appsv1.Deployment, bool) {
+	pod, ok := builder.makePodForContainer(dbc)
```
Btw, one thing that you could do to make it a bit easier to unit test, is pass the pod into makeDeploymentForContainer rather than generating it within.  Not sure if it will clean things up much or not, just a thought.

ejj commented on minion/kubernetes/deployment.go:
```diff
@@ -0,0 +1,405 @@
+	return deploymentBuilder{imageMap, toAffinities(placements), secretClient}
+}
+
+func (builder deploymentBuilder) makeDeploymentForContainer(dbc db.Container) (
```
Also, I think `makeDeployment` and `makePod` are fine names.  The ForContainer isn't really giving any additional information.  Particular since it takes a container as an argument.

ejj commented on minion/kubernetes/deployment.go:
```diff
@@ -0,0 +1,405 @@
+	return envVars
+}
+
+func toPodVolumes(filepathToContent map[string]blueprint.ContainerValue) (
```
I might call these `toPod` functions `makePod` instead.  Would be more consistent with the `makePod()` function above.

ejj commented on minion/kubernetes/deployment.go:
```diff
@@ -0,0 +1,405 @@
+	return
+}
+
+func hashStr(toHash string) string {
```
Btw, how securely does kubernetes treat environment variables?  Leaking sha1s of secrets isn't ideal ...  Not sure what to do about it, but thought I'd mention it for the record.

ejj commented on minion/kubernetes/status.go:
```diff
@@ -0,0 +1,176 @@
+		return joinKey{
+			Hostname: dbc.Hostname,
+			Image:    dbc.Image,
+			Command:  fmt.Sprintf("%+v", dbc.Command),
```
It's a bit odd that this one uses `%+v` and the one below simply uses `%v`.  I don't think it should matter, but was that intentional?

ejj commented on minion/network/openflow/run.go:
```diff
@@ -0,0 +1,95 @@
+	log "github.com/sirupsen/logrus"
+)
+
+// Run occaisionally updates the OpenFlow tables based on the containers in the database.
```
occasionally is misspelled

kklin replied:
Haha, that's actually from your PR. But I'll fix

kklin commented:
@ejj I addressed your comments, and implemented the volume/privileged container stuff. The commits with new code are:
- minion: Use blueprint row in database 
- kube: Squash me
- vendor: Remove unused dependencies

ejj commented:
@kklin Looked at the new patches, all good.  Merge when ready.
